api/api/utils/image_proxy/__init__.py

```diff
@@ -88,9 +122,10 @@ def get(
 
     logger = parent_logger.getChild("get")
     tallies = django_redis.get_redis_connection("tallies")
+    tallies_incr = sync_to_async(tallies.incr)
```

Comment:
Sorry I'm missing a lot of sync-to-async context since this is the first such PR that I'm reviewing. Why do we not use `async def`, but  need an `sync_to_async` for tallies?

Reply:
We are using `async def`, unless I unintentionally missed it somewhere. `async def` is only necessary when awaiting, though.

`sync_to_async` is necessary to prevent context issues in the Redis client, which assumes that only one thread is every touching the connection pool. `sync_to_async` creates a safe context for such synchronous functions. Relevant to Django ORM as well. [From Django's documentation](https://django.readthedocs.io/en/stable/topics/async.html#asgiref.sync.sync_to_async):

> The reason this is needed in Django is that many libraries, specifically database adapters, require that they are accessed in the same thread that they were created in. Also a lot of existing Django code assumes it all runs in the same thread, e.g. middleware adding things to a request for later use in views.

As an additional point of clarification, Django's own async cache support just wraps the synchronous versions of functions with `sync_to_async`, rather than using fully async clients. In the future we could switch to an asynchronous Redis client (like https://github.com/Andrew-Chen-Wang/django-async-redis, but I'm not sure that it's 100% stable), but for now it's not necessary and Django's own caching behaviour would still use the `sync_to_async` adapter approach for this anyway.

So the reason here is to create a new async version of `tallies.incr` that we can call later with `await`.

Reply:
Let me know if I answered your question or if I misunderstood. The Django documentation on async adaption is a good read and important for understanding how and when it's necessary to adapt synchronous functions to an async context.